core/capabilities/ccip/oraclecreator/bootstrap.go

```diff
@@ -104,5 +133,241 @@ func (i *bootstrapOracleCreator) Create(_ uint32, config cctypes.OCR3ConfigWithM
+// Required for managing RMN related peer group connections.
+type peerGroupDialer struct {
+	lggr logger.Logger
+
+	peerGroupFactory rmn.PeerGroupFactory
+	rmnHomeReader    ccipreaderpkg.RMNHome
+
+	// common oracle config
+	bootstrapLocators  []commontypes.BootstrapperLocator
+	oraclePeerIDs      []ragep2ptypes.PeerID
+	commitConfigDigest [32]byte
+
+	activePeerGroups    []rmn.PeerGroup
+	activeConfigDigests []cciptypes.Bytes32
+
+	syncInterval time.Duration
+
+	mu        *sync.Mutex
+	syncCtxCf context.CancelFunc
+}
+
+func newPeerGroupDialer(
+	lggr logger.Logger,
+	peerGroupFactory rmn.PeerGroupFactory,
+	rmnHomeReader ccipreaderpkg.RMNHome,
+	bootstrapLocators []commontypes.BootstrapperLocator,
+	oraclePeerIDs []ragep2ptypes.PeerID,
+	commitConfigDigest [32]byte,
+) *peerGroupDialer {
+	return &peerGroupDialer{
+		lggr: lggr,
+
+		peerGroupFactory: peerGroupFactory,
+		rmnHomeReader:    rmnHomeReader,
+
+		bootstrapLocators:  bootstrapLocators,
+		oraclePeerIDs:      oraclePeerIDs,
+		commitConfigDigest: commitConfigDigest,
+
+		activePeerGroups: []rmn.PeerGroup{},
+
+		syncInterval: time.Minute, // todo: make it configurable
+
+		mu:        &sync.Mutex{},
+		syncCtxCf: nil,
+	}
+}
+
+func (d *peerGroupDialer) Start() {
+	if d.syncCtxCf != nil {
+		d.lggr.Warnw("peer group dialer already started, should not be called twice")
+		return
+	}
+
+	d.lggr.Infow("Starting peer group dialer")
+
+	ctx, cf := context.WithCancel(context.Background())
+	d.syncCtxCf = cf
+
+	go func() {
+		d.sync()
+
+		syncTicker := time.NewTicker(d.syncInterval)
+		for {
+			select {
+			case <-syncTicker.C:
+				d.sync()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (d *peerGroupDialer) Close() error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	d.closeExistingPeerGroups()
+
+	if d.syncCtxCf != nil {
+		d.syncCtxCf()
+	}
+
+	return nil
+}
+
+func (d *peerGroupDialer) sync() {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if !d.shouldSync() {
+		d.lggr.Debugw("No need to sync peer groups")
+		return
+	}
+	d.lggr.Infow("Syncing peer groups")
+
+	d.closeExistingPeerGroups()
+
+	if err := d.createNewPeerGroups(); err != nil {
+		d.lggr.Errorw("failed to create new peer groups", "err", err)
+		d.closeExistingPeerGroups() // close potentially opened peer groups
+	}
+}
+
+func (d *peerGroupDialer) shouldSync() bool {
+	if len(d.activePeerGroups) == 0 {
+		return true
+	}
+
+	activeConfigDigest, candidateConfigDigest := d.rmnHomeReader.GetAllConfigDigests()
+	var configDigests [][32]byte
+
+	if !activeConfigDigest.IsEmpty() {
+		configDigests = append(configDigests, activeConfigDigest)
+	}
+	if !candidateConfigDigest.IsEmpty() {
+		configDigests = append(configDigests, candidateConfigDigest)
+	}
+
+	if len(configDigests) != len(d.activeConfigDigests) {
+		return true
+	}
+	for i, rmnHomeConfigDigest := range configDigests {
+		if rmnHomeConfigDigest != d.activeConfigDigests[i] {
+			return true
+		}
+	}
```

Comment:
I think I see several issues with this code, here's just one of them:

* if current state is [active, candidate] and both are not 0x0:
* if onchain state is [active, candidate] and candidate is 0x0:
* then we won't spin down the candidate peer group, because `configDigests` will be of length 1, and the loop on line 287 will only check the active digest.

My recommendation would be for this func to return maybe an array of "actions" to do - e.g spin down X, spin up Y, etc.